syntax = "proto2";
package gt.stinger;

import "stinger_net/proto/stinger-batch.proto";

enum AlgAction {
  REGISTER_ALG	      = 0;
  BEGIN_INIT	      = 1;
  END_INIT	      =	2;
  BEGIN_PREPROCESS    = 3;
  END_PREPROCESS      = 4;
  BEGIN_POSTPROCESS   = 5;
  END_POSTPROCESS     = 6;
}

enum AlgActionResult {
  ALG_SUCCESS			      = 0;
  ALG_FAILURE_GENERIC     	      = 1;
  ALG_FAILURE_NAME_EXISTS 	      = 2;
  ALG_FAILURE_NAME_INVALID	      = 3;
  ALG_FAILURE_DEPENDENCY  	      = 4;
  ALG_FAILURE_UNEXPECTED_MESSAGE      = 5;
  ALG_FAILURE_TIMEOUT		      =	6;
}

message AlgToServer {
  optional int64      protocol_version	= 1 [default = 1];

  optional string     alg_name		= 2;
  optional int64      alg_num		= 3;
  optional int64      data_per_vertex	= 4 [default = 0];
  optional string     data_description	= 5;
  required AlgAction  action		= 6;
  repeated string     req_dep_name	= 7;
  repeated string     opt_dep_name	= 8;
}

message ServerToAlg {
  optional int64	    protocol_version	= 1 [default = 1];

  optional string	    alg_name	      	= 2;
  optional int64	    alg_num	      	= 3;
  required AlgAction	    action	      	= 4;
  required AlgActionResult  result	      	= 5;
  optional int64	    batch_number      	= 6 [default = -1];

  optional StingerBatch	    batch	      	= 7;
  optional string	    stinger_loc	      	= 8;
  optional int64	    stinger_size	= 9;
  optional string	    alg_data_loc      	= 10;
  repeated string	    dep_name		= 11;
  repeated string	    dep_description	= 12;
  repeated string	    dep_data_loc      	= 13;
  repeated int64	    dep_data_per_vertex	= 14;
}
